### DISCUSSION TITLE
Splitting arbitrary arrays into arbitrary chunks

### DISCUSSION THREAD

**Discussion by skybrian** [dubzzz/fast-check#2916] [Ideas]
Hi, I'm new to fast-check and I'm testing some code that assembles a stream of binary chunks of UTF8-encoded strings into JavaScript strings. The way I test this (for the case of valid data) is by choosing an arbitrary unicode string and a number indicating a split point, and then converting the string to binary and splitting it at that point:

```typescript
function byteStream(input: fc.Arbitrary<string>): fc.Arbitrary<[string, ReadableStream<Uint8Array>]> {
  const splits =
    fc.tuple(input, fc.nat())
      .map(([s, n]) => [s, new TextEncoder().encode(s), n] as const)
      .filter(([_, b, n]) => n <= b.length);

  return splits.map(([s, b, n]) => {
    const stream = new ReadableStream({
      start: (controller) => {
        controller.enqueue(b.slice(0, n));
        controller.enqueue(b.slice(n));
        controller.close();
      }
    });
    return [s, stream];
  });
}
```

That only handles the case of two chunks, and I'd like to expand testing to an arbitrary number of chunks. It seems like I need more random numbers for that. I could generate an array of natural numbers for chunk sizes and filter it out if they sum to longer than the binary data, but that seems wasteful?

I'm wondering if the case where you want to split arbitrary arrays into arbitrary chunks is common enough that fast-check should have a built-in way to do it. Or maybe there's a better way to go about this?

**Comment by tskj** [comment]
If I understand correctly, I think this case could be solved pretty simply by generating an array of natural numbers (as you suggest), but consider them indices into the binary data, and use those indices as chunk slices. Of course also use the modulo of each natural number, as in `n % stream.length` to get valid indices.

**Reply by skybrian** [reply]
Thanks!
